af_ktls.c

```diff
@@ -984,10 +988,14 @@ static inline ssize_t tls_peek_data(struct tls_sock *tsk, unsigned flags)
 			goto peek_failure;
 		}
 
+		if (peeked_size < KTLS_MIN_RECORD_SIZE(tsk)) {
+			ret = -EBADF;
```

Comment:
> When peeked size is bigger than header: Only peek based on the header size

I wanted to avoid the overhead of doing peek twice just to get actual records size. Now I am thinking whether we can actually avoid it.

> Anything larger is a kernel bug. This shouldn't ever happen otherwise BUG()

Can we rely on this? What if the record was malformed (but checksum passed)?

> For TCP
>
> When peeked size is smaller than header: We just haven't received enough data yet. We should signal EAGAIN to userspace if it is a nonblocking socket.
> If it is a blocking socket, the thread should go to sleep until we get another sk_data_ready and do peek the right amount of data.
>
> For UDP Yes I think we can signal an error for peeked size to small, I believe sk_data_ready isn't called unless we have a full datagram

I agree.